Return wpcom request result from post action dispatch

jsnajdr opened the pull request:
This is a spinoff PR from #23090 with a part of the work that can be reviewed and merged independently. It improves the posts action thunks in a few ways:

**Return wpcom request result from post action dispatch**

This patch changes the `savePost`, `deletePost` and `restorePost` actions to return the `wpcom` request result from the dispatch. It allows us to chain promise handlers to the dispatch call and set local state in them:
```js
dispatch( restorePost( ... ) )
  .then( restoredPost => this.handleRestoreSuccess( restoredPost ) )
  .catch( error => this.handleRestoreError( error ) );
```

**Update the post action tests to cope with rejected promises**

The action dispatches now return both resolved and rejected promises. Update the tests so that they can cope also with rejected promises. We do that by inserting a dummy `catch( noop )` into the handler chain. An ideal solution would be to use `Promise.prototype.finally`, but that is a very new Promise feature that's not everywhere yet.

**Use wpcom.site().post().delete() to trash a post**

Trashing post is almost equivalent to saving the post with status field set to `trash` and the action behaves like it was doing exactly that -- it dispatches the `POST_SAVE_*` actions with the right properties.

But what we really do is to call the `wpcom.site().post().delete()` method, i.e., sending a `POST /sites/:site/posts/:post/delete` request. The difference is that an explicit delete will set a `_wp_trash_meta_status` meta property on the post and a later `restore` call can restore the original post status, i.e., `publish`. Without this, the post will be always recovered as `draft`.

**How to test:**
Only the changes in how a post is trashed affect the UI:
1. Go to the list of published posts or pages
2. Trash a published post or page
3. Go to the list of trashed items and restore the item you just trashed

Before this patch: the post/page is always recovered as "Draft"
After this patch: the post/page is recovered as "Published"

Similarly, recovering an item that was previously "Draft" or "Scheduled" will recover it with the original status.

jsnajdr committed:
Update the post action tests to cope with rejected promises

The action dispatches now return both resolved and rejected promises. Update the tests
so that they can cope also with rejected promises. We do that by inserting a dummy
`catch( noop )` into the handler chain. An ideal solution would be to use
`Promise.prototype.finally`, but that is a very new Promise feature that's not everywhere yet.

jsnajdr committed:
Return wpcom request result from post action dispatch

This patch changes the `savePost`, `deletePost` and `restorePost` actions to return
the `wpcom` request result from the dispatch. It allows us to chain promise handlers
to the dispatch call and set local state in them:
```
dispatch( restorePost( ... ) )
  .then( restoredPost => this.handleRestoreSuccess( restoredPost ) )
  .catch( error => this.handleRestoreError( error ) );
```

jsnajdr committed:
Use wpcom.site().post().delete() to trash a post

Trashing post is almost equivalent to saving the post with status field set to `trash`
and the action behaves like it was doing exactly that -- it dispatches the `POST_SAVE_*`
actions with the right properties.

But what we really do is to call the `wpcom.site().post().delete()` method, i.e., sending
a `POST /sites/:site/posts/:post/delete` request. The difference is that an explicit delete
will set a `_wp_trash_meta_status` meta property on the post and a later `restore` call
can restore the original post status, i.e., `publish`. Without this, the post will be always
recovered as `draft`.

nylen commented:
> Return wpcom request result from post action dispatch. It allows us to chain promise handlers to the dispatch call and set local state in them

I think this is probably the most expedient way to achieve the goals of #23090, but as I noted a bit in https://github.com/Automattic/wp-calypso/pull/23090#issuecomment-374296812 I am not totally sold on this approach.  It's adding another place where we need to look for side effects that happen after certain actions.

> Update the post action tests to cope with rejected promises [using `.catch( noop )`, because...] action dispatches now return both resolved and rejected promises.

My problem with this approach is that if a test is written like `.catch( noop ).then( expect( ... ) )`, then the test will pass regardless of whether the promise was rejected or resolved.

I tried to test the expected cases more explicitly (test rejected promises in `catch` only, and resolved promises in `then` only).  I also wanted to use `done` to ensure that the test's assertions are actually called as expected.  However, I ran into some unexpected behavior that I'll illustrate in vanilla JS:

```js
const p = new Promise( () => { throw 'broken'; } );
p.catch( () => console.log( 'error 1' ) );
p.catch( () => console.log( 'error 2' ) );
```

Result: The two `catch` callbacks are executed in the order specified, as expected.

```js
const p = new Promise( () => { throw 'broken'; } );
p.then( () => console.log( 'ok' ) ).catch( () => console.log( 'error 1' ) );
p.catch( () => console.log( 'error 2' ) );
```

Result: The second `catch` callback is executed **before** the first `catch` callback.

In this particular case, this means that any `catch` handlers specified on the return value of `savePost` will be executed **before** the `catch` handler within `savePost` itself.  This doesn't seem desirable, and it makes it impossible to write these tests the way I think they should be written.  For example, this doesn't work because the `expect` happens **before** the `POST_SAVE_FAILURE` dispatch:

```js
test( 'should dispatch failure action when error occurs while saving existing post', done => {
    return savePost( 77203074, 102, { title: 'Hello World' } )( spy )
        .catch( () => {
            expect( spy ).to.have.been.calledWith( {
                type: POST_SAVE_FAILURE,
                siteId: 77203074,
                postId: 102,
                error: sinon.match( { message: 'User cannot edit post' } ),
            } );
            done();
        } );
} );
```

nylen committed:
Explicitly test whether promises were resolved or rejected

Also use done() to ensure that assertions are called as expected.

nylen commented:
558513d is more like what I had in mind for the tests.  Now, if any of the promises are resolved when they are expected to be rejected (or vice versa), the tests will fail.  Generally this was not guaranteed to happen before.

I don't love the `setTimeout` calls but that is a workaround for the unexpected ordering of `catch` calls described above.

jsnajdr committed:
Post Actions: ensure that outside promise handlers are called after the internal ones

nylen committed:
Only include `done` where necessary

For tests that perform assertions inside `.catch` blocks, we need to
call `done()` to ensure that the promise being tested was actually
rejected.

For other tests, returning a promise is recognized by Jest as a way to
mark a test as async, so `done` is not needed.

nylen committed:
Fix lint error

jsnajdr commented:
@nylen The funny order of the promise callback is a great catch, thanks! It's actually a bug in the tested code -- the outside handlers should be called only after the internal ones, i.e., the ones that dispatch Redux actions.

How is the order determined?
```js
p.then( onSuccessIn ).catch( onErrorIn );
p.catch( onErrorOut );
```

Promise `p` has two handlers attached, one is `then`, one is `catch`. When the promise is rejected, both of them are scheduled to be processed in the next event loop tick.

The `then` handler doesn't execute the `onSuccessIn` handler, but creates a new rejected promise and the `catch( onErrorIn )` handler is scheduled for the next event loop tick.

The `catch` handler executes the `onErrorOut` handler and uses its return value to create a new promise. That new promise is garbage collected, because the return value is not used.

Only after both above handlers are done, the `catch( onErrorIn )` is executed. That's the behavior you've been seeing.

There are two solutions:
```js
p.then( onSuccessIn, onErrorIn );
p.catch( onErrorOut );
```
No intermediate promise is created here, all handlers are attached directly to `p`. They will be executed in the same order in which they got attached, i.e., `onSuccessIn`, `onErrorIn`, `onErrorOut`.

Second solution is:
```js
const q = p.then( res => {
  onSuccessIn( res );
  return res;
} ).catch( err => {
  onErrorIn( err );
  throw err; // or return Promise.reject( err )
} );
q.catch( onErrorOut );
```
The order of execution is the same as with Solution 1.

Both solutions behave differently when `onSuccessIn` throws an exception: `onErrorIn` handles it in one case and doesn't in the other.

I ended up using Solution 1, i.e., `then` handler with two callbacks.

nylen reviewed (approved):
I have my doubts about whether we should make extensive use of this technique (use promises to pull side effects into individual components), but this change is good to go.  As written, the tests ensure that callback order is as expected.  :+1:

jsnajdr commented:
> I have my doubts about whether we should make extensive use of this technique (use promises to pull side effects into individual components)

For me it's the No.1 most wanted feature I'd like to see in all Calypso actions, including data layer :smile: I believe many UI interactions are simply impossible to implement without something like this.

For example, what if we wanted to implement two different UIs over the same post list query? If everything must be in Redux, the only UI we are stuck with are global notices.

Another case where I needed an action dispatch to return a Promise was when creating a Simple Payment and then inserting a shortcode with its ID into a TinyMCE editor window. As TinyMCE UI is not React/Redux based, I need to write something like:
```js
dispatch( createSimplePayment() ).then(
  payment => this.tinymce.insertText( shortcodeString( payment.ID ) )
);
```

Returning promises from action dispatches and composing them to perform complex async logic is quite common in Firefox codebase, e.g., in the [debugger frontend](https://github.com/devtools-html/debugger.html).

nylen commented:
Where to handle side effects is an interesting discussion.  I can think of a few things I've done or seen recently that would have been easier to implement using an action creator that returns a promise ([example](https://github.com/Automattic/wp-calypso/pull/22111/files#diff-1aa09cd6ba66c33617a1d7b3a4f95f06R40)), but I still don't think it's the right solution for all cases (if core application logic ends up being handled this way, then it seems like it will become pretty messy).

> what if we wanted to implement two different UIs over the same post list query? If everything must be in Redux, the only UI we are stuck with are global notices.

We would be stuck with UI patterns that fit well within the global application state.  If a behavior pattern is really a core behavior of the data of the application, or even just something that will end up being implemented in multiple components, then it probably makes more sense for it to go in the Redux state and we should use Redux mechanisms to drive the UI changes around that behavior.  For other things I guess it depends, sometimes Redux will end up being cleaner and sometimes it won't.
